02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  IChargingStationSecurityInfoRepository,
+  IDeviceModelRepository,
+  sequelize,
+} from '@citrineos/data';
+import {
+  IFileAccess,
+  MeterValueType,
+  SignedMeterValuesConfig,
+  SignedMeterValueType,
+  SystemConfig,
+} from '@citrineos/base';
+import { ILogObj, Logger } from 'tslog';
+import * as crypto from 'node:crypto';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const expectPublicKey =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    for (const meterValue of meterValues) {
+      if (expectPublicKey.length > 0 && expectPublicKey[0].value !== 'Never') {
```

Comment:
Going off of this, can also put the stuff that the for loop executes into its own method. Then, this method can look like:
```
  public async validateMeterValues(
    stationId: string,
    meterValues: [MeterValueType, ...MeterValueType[]],
  ): Promise<boolean> {
    let validMeterValues = true;

    const expectPublicKey =
      await this._deviceModelRepository.readAllByQuerystring({
        stationId,
        component_name: 'OCPPCommCtrlr',
        variable_name: 'PublicKeyWithSignedMeterValue',
      });

    if (expectPublicKey.length < 0 || expectPublicKey[0].value === 'Never') {
      return true;
    }

    for (const meterValue of meterValues) {
      for (const sampledValue of meterValue.sampledValue) {
        const signedMeterValue = sampledValue.signedMeterValue;
        if (signedMeterValue) {
          if (!this.validateSignedMeterValue(stationId, signedMeterValue)) {
            return false;
          }
        }
      }
    }

    return true;
  }
```

Although I think you'd want to update such that when

Reply:
i left it in the loop on the crazy chance that someone sent multiple meter values, but not all of them were signed -- i don't think that's possible, right?? either all meter values are signed or they're not 😆 i'll update and refactor 👍

Reply:
refactored so that `expectPublicKey` is renamed to `shouldPublicKeyBeUsedForValidation` for clarity + moving it outside the for loop + refactored the logic into a helper method for readability 👍